Fix fill factor, error tightening and growing ratio.

arthurprs opened the pull request:
Hello, I was using the crate in a project so I spotted and addressed a few shortcomings.

Problem: BitVec Serialization format is sub-optimal, which might not be good when a bloom filter is involved.
Solution: Use a simple Vec<u8> and serialize as bytes.

Problem: Empty filter allocates
Solution: Don't eagerly add a filter on new()

Problem: Grow and error tightening factor wasn't clear. It was sort-of adding 1 slice per growth?
Solution: Use a 2x growth factor and 0.8 tightening factor, which gives a ~5x upper bound on fp ratio even after 1000x growth.

Problem: Arbitrary hashing strategy
Solution: Use enhanced-double-hashing strategy which is both simpler and mathematically correct.

Problem: Bit fill ratio of 0.8 results in a completely off false positive ratio. BFs are built around a 0.5 fill factor. In addition, that's not practical performance wise.
Solution: Use a insert counter to estimate fill factor, which is sort-of the standard for performant implementations.

Not correctness related:
* Switched to ahash hasher, which became the de-facto performance focused hasher in the ecosystem.

dpbriggs commented:
Hey! Thank you for the pull request, this is awesome. 

I'll be at work for the next ~8 hours, and then I'll be able to review properly. First a few ideas:

- This will need to be a major version bump; could you change the version to 2.0.0?
- Changing the struct layout will likely break existing uses of the library. We should add some documentation near the top of the README. If you can modify the org-mode that would be best, then export. Otherwise I could do that, then export to markdown, then remove the org mode file
- Would it be desirable to allow users to choose the seed?
- Would it be desirable to allow users to choose the hashing backend? (default to double hashing; trait + empty struct?)
- And would you mind adding some more documentation around the `TIGHTENING_RATIO`? Was that also referenced in the paper?
- I don't mind changing the grow logic to be based on the true number of inserts, but when users insert a lot of items that hash similarly we may start growing the filter unnecessarily according to the false positive rate. Do you know if this strategy is popular with other bloom filters? I pulled the `fill_ratio_gte` from the paper

dpbriggs commented on src/lib.rs:
```diff
@@ -119,64 +114,29 @@ impl Bloom {
+    // h2 hash shouldn't be 0 for double hashing
+    let h2 = hs.finish().max(1);
+
+    (h1 as _, h2 as _)
```
It's kinda ugly but a common convention in my code bases is to not use implicit conversions like this. It's a-ok here, so read-and-resolve

dpbriggs commented on src/lib.rs:
```diff
@@ -313,19 +267,20 @@ impl GrowableBloom {
     /// bloom.insert(&vec![1, 2, 3]);
```
Could we update the doc comment to match the new API? And definitely clarify in the doc comment that if you get true, your item definitely wasn't in the filter (it's easy to confuse this)

dpbriggs commented on src/lib.rs:
```diff
@@ -369,20 +337,17 @@ impl GrowableBloom {
-        let new_bloom = Bloom::new(self.curr_num_slice, self.slice_size, self.curr_seed);
-        self.blooms.push(new_bloom)
+        let error_ratio =
+            self.desired_error_prob * Self::TIGHTENING_RATIO.powi(self.blooms.len() as _);
```
Would you mind leaving a comment explaining the idea behind this? You have a comment on one of the tests that you could re-use. It's been a long time since I read the paper, and it'll probably confuse newcomers as well

dpbriggs reviewed (changes requested):
Hey! Sorry for not getting back to you last night. It might be looking like I'm asking for a lot of changes but most of these review comments are asking for comments/documentation updates.

If you're busy I don't mind commenting/documenting the hell out of this, so we can get it merged.

Overall I think you did a great job, and am happy with the changes. I'm excited to merge this

& I need to figure out how to actually checkout your branch so I can do a final export of the org-mode file to markdown, then remove the org-mode file. I'll push changes

arthurprs commented on src/lib.rs:
```diff
@@ -16,69 +16,64 @@ use std::{
 
-    /// Insert an `item` into the Bloom.
+    /// Insert an *NEW* `item` into the Bloom.
+    /// The caller is expected to check that the item is not contained
```
Ahh, this comment change is not needed anymore. A leftover from a previous iteration.

arthurprs commented on src/lib.rs:
```diff
@@ -16,69 +16,64 @@ use std::{
+            let hi = h1 % slice_len + i * slice_len;
+            h1 = h1.wrapping_add(h2);
+            h2 = h2.wrapping_add(i);
+            (hi / 8, 1 << (hi % 8))
```
:+1:

arthurprs commented:
I think I addressed all your comments with the last couple comments. But I realized that there're some problems with regards to determinism across platforms (eg. endianness changes and 32/64 changes), which is not a good thing when the filter is supposed to be (de)serializable.

I'll address that now.

dpbriggs commented on src/lib.rs:
```diff
@@ -22,35 +23,39 @@ struct Bloom {
     fn new(capacity: usize, error_ratio: f64) -> Bloom {
+        // Directly from paper:
+        // k = log2(1/P)   (num_slices)
+        // n ≈ −m ln(1−p)  (slice_len_bits)
```
I'm a little spooked about this approximation, as the paper says 'For usable values of _m_` - are our slice lengths really that large? The paper says: "For any given error probability _P_ and filter size _M_, _n_ is maximized by making _p= 1/2_, regardless of _P_ or _M_."

I think the old formula is better, and I vaguely recall other libraries using the old formula

arthurprs replied:
This is tricky. It's important to consider that all formulas are approximations. The problem with the previous approach/formula is that it's derived from the optimal value of K, which is unrealistic since K _must_ be an integer. The end result is that the resulting filter would have a slightly off false-positive ratio.

The way I have it now will at least guarantee we get a BF with a false-positive rate <= the desired one, which is important to get a correct "upper-bound" when growing the scalable bloom filter.

Bloom filters are incredibly easy to get "slightly wrong" honestly. I'm doing my best to get this one right.

dpbriggs replied:
Sounds good, and that's fair. I'm pretty comfortable with the changes at this stage. I'm not sure if you're still experimenting/making changes, so lemme know when you're comfortable with the changes

arthurprs replied:
I think it's good to go, lets leave it open for a few days just in case.

arthurprs replied:
I just realized that we also need something like this https://github.com/rust-lang/rust/blob/c0955a34bcb17f0b31d7b86522a520ebe7fa93ac/src/librustc_data_structures/stable_hasher.rs#L78-L166 to wrap the hasher. Otherwise hashing integers will have different outputs in LE/BE systems.

dpbriggs commented on src/lib.rs:
```diff
@@ -22,35 +23,39 @@ struct Bloom {
+        // Directly from paper:
+        // k = log2(1/P)   (num_slices)
+        // n ≈ −m ln(1−p)  (slice_len_bits)
+        // M = k * n       (total_bits)
```
I think `n` and `m` got mixed up here

dpbriggs commented on src/lib.rs:
```diff
@@ -22,35 +23,39 @@ struct Bloom {
         Bloom {
             buffer: buffer.into_boxed_slice(),
-            num_slices,
+            num_slices: NonZeroU64::new(num_slices).unwrap(),
```
I really like this, nice :+1:

dpbriggs commented on src/lib.rs:
```diff
@@ -157,19 +165,15 @@ impl Bloom {
+    // which we know is the case.
     0u8.hash(&mut hasher);
     // h2 hash shouldn't be 0 for double hashing
     let h2 = hasher.finish().max(1);
```
Could we leave a comment explaining why the first one doesn't need `.max(1)`?

arthurprs replied:
I don't know how to word that. It's easy to comment when something is a special (h1) but not obvious otherwise. The linked Double hashing article should be enough honestly.

dpbriggs replied:
Sounds good :+1:

dpbriggs reviewed (changes requested):
Things are trending a great direction - thank you so much for the work so far.

What are your thoughts on the slice-length calculation? The paper gives a few options, and I'm preferring the original one. What do you think?

dpbriggs commented on src/stable_hasher.rs:
```diff
@@ -0,0 +1,99 @@
+    }
+}
+
+impl Hasher for StableHasher {
```
This is fantastic stuff. How are you feeling about this PR?

arthurprs replied:
I think it might be good to go now :+1:

arthurprs replied:
I may have one more change. I'll push shortly.

dpbriggs replied:
I just reviewed the changes and things are looking great. You comfortable with me merging?

arthurprs replied:
:shipit: !

dpbriggs replied:
I also just tested the benchmarks and the numbers are looking way better. Suprisingly consistent across different failure rates, which is a great sign.

arthurprs replied:
Awesome. Could you make a crates.io release with this new version?

dpbriggs replied:
For sure. Just got off work so I'll merge and prep for all of that

dpbriggs commented:
@arthurprs [Published 2.0.0 ](https://crates.io/crates/growable-bloom-filter) :rocket:

 Thank you so much for your incredible contribution! I really appreciated your patience, and it was neat going through my first real pull request. You have a great one!
